File: year2025/day7/part1/solution.py
#!/usr/bin/env python3
# Day 7: Laboratories, Part 1

FILE_PATH = "../.input"

SOURCE = 'S'
TACHYON_BEAM = '|'
EMPTY_SPACE = '.'
SPLITTER = '^'

def print_manifold(manifold: list[list[int]]) -> None:
    print(f"\n{'\n'.join(''.join(i) for i in manifold)}")

def main():
    split_count = 0
    manifold = []

    with open(FILE_PATH, 'r') as manifold_file:
        for line in manifold_file:
            manifold.append(list(line.strip()))

    #print("DEBUG: starting manifold: ", end='')
    #print_manifold(manifold)

    for j in range(num_cols):
        if manifold[0][j] == SOURCE:
            manifold[0][j] = TACHYON_BEAM
            break

    for i, row in enumerate(manifold):
        #print(f"DEBUG: row {i:03}: {''.join(row)}")
        if i == len(manifold) - 1:
            break
        for j, char in enumerate(row):
            if char == TACHYON_BEAM:
                below = manifold[i+1][j]
                if below == EMPTY_SPACE:
                    manifold[i+1][j] = TACHYON_BEAM
                elif below == SPLITTER:
                    split_count += 1
                    if manifold[i+1][j-1] == EMPTY_SPACE:  # Left
                        manifold[i+1][j-1] = TACHYON_BEAM
                    if manifold[i+1][j+1] == EMPTY_SPACE:  # Right
                        manifold[i+1][j+1] = TACHYON_BEAM

    #print("DEBUG: ending manifold: ", end='')
    #print_manifold(manifold)

    print(f"Total splits: {split_count}")

if __name__ == "__main__":
    main()

File: year2025/day7/part1/solution.cpp
// Day 7: Laboratories, Part 1

#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

#define SOURCE 'S'
#define TACHYON_BEAM '|'
#define EMPTY_SPACE '.'
#define SPLITTER '^'

static const string FILE_PATH = "../.input";

void print_manifold(const vector<vector<char>>& manifold) {
    for (const auto& row : manifold) {
        for (const char c : row) {
            cout << c;
        }
        cout << '\n';
    }
}

int main() {
    unsigned long long split_count = 0;
    vector<vector<char>> manifold;

    ifstream manifold_file(FILE_PATH);
    string line;
    while (getline(manifold_file, line)) {
        manifold.emplace_back(line.begin(), line.end());
    }
    manifold_file.close();

    //cout << "DEBUG: starting manifold:\n";
    //print_manifold(manifold);

    for (char& c : manifold[0]) {
        if (c == SOURCE) {
            c = TACHYON_BEAM;
            break;
        }
    }

    for (size_t i = 0; i < manifold.size() - 1; i++) {
        for (size_t j = 0; j < manifold[i].size(); j++) {
            //cout << "DEBUG: manifold[" << i << "][" << j << "]: " << manifold[i][j] << "\n";
            if (manifold[i][j] == TACHYON_BEAM) {
                if (manifold[i+1][j] == EMPTY_SPACE) {
                    manifold[i+1][j] = TACHYON_BEAM;
                } else if (manifold[i+1][j] == SPLITTER) {
                    split_count++;
                    if (manifold[i+1][j-1] == EMPTY_SPACE) {
                        manifold[i+1][j-1] = TACHYON_BEAM;
                    }
                    if (manifold[i+1][j+1] == EMPTY_SPACE) {
                        manifold[i+1][j+1] = TACHYON_BEAM;
                    }
                }
            }
        }
    }

    //cout << "DEBUG: ending manifold:\n";
    //print_manifold(manifold);

    cout << "Total splits: " << split_count << endl;

    return 0;
}
